esrally/storage/_adapter.py

```diff
@@ -0,0 +1,154 @@
+    # Example:
+    #   ```
+    #   class HTTPAdapter(Adapter):
+    #       # The value will serve to associate any URL with "https" scheme to `HTTPAdapter` subclass.
+    #       __adapter_URL_prefixes__ = ("http://",)
+    #   ```
+    __adapter_URL_prefixes__: tuple[str, ...] = tuple()
+
+    @classmethod
+    def from_config(cls, cfg: Config) -> Adapter:
+        return cls()
+
+    @abstractmethod
+    def head(self, url: str) -> Head:
+        """It gets remote file headers.
+        :return: the Head of the remote file.
+        :raises ServiceUnavailableError: in case on temporary service failure.
+        """
+
+    @abstractmethod
+    def get(self, url: str, stream: Writable, ranges: RangeSet = NO_RANGE) -> Head:
+        """It downloads a remote bucket object to a local file path.
+
+        :param url: it represents the URL of the remote file object.
+        :param stream: it represents the local file stream where to write data to.
+        :param ranges: it represents the portion of the file to transfer (it must be empty or a continuous range).
+        :raises ServiceUnavailableError: in case on temporary service failure.
+        """
+
+
+ADAPTER_CLASS_NAMES = ",".join(
+    [
+        "esrally.storage._http:HTTPAdapter",
+    ]
+)
+
+
+class AdapterClassEntry(NamedTuple):
+    url_prefix: str
+    cls: type[Adapter]
+
+
+class AdapterRegistry:
+    """AdapterClassRegistry allows to register classes of adapters to be selected according to the target URL."""
+
+    def __init__(self, cfg: Config) -> None:
+        self._classes: list[AdapterClassEntry] = []
+        self._adapters: dict[type[Adapter], Adapter] = {}
+        self._lock = threading.Lock()
+        self._cfg = cfg
+
+    @classmethod
+    def from_config(cls, cfg: Config) -> AdapterRegistry:
+        registry = cls(cfg)
+        adapters_specs = (
+            cfg.opts(section="storage", key="storage.adapters", default_value=ADAPTER_CLASS_NAMES, mandatory=False)
+            .replace(" ", "")
+            .split(",")
+        )
+        for spec in adapters_specs:
+            module_name, class_name = spec.split(":")
+            module = importlib.import_module(module_name)
+            obj = getattr(module, class_name)
+            if not isinstance(obj, type) or not issubclass(obj, Adapter):
+                raise TypeError(f"'{obj}' is not a valid subclass of Adapter")
+            registry.register_class(obj, obj.__adapter_URL_prefixes__)
+        return registry
+
+    def register_class(self, cls: type[Adapter], prefixes: Iterable[str]) -> type[Adapter]:
+        with self._lock:
+            for p in prefixes:
+                self._classes.append(AdapterClassEntry(p, cls))
+            # The list of adapter classes is kept sorted from the longest prefix to the shorter to ensure that matching
+            # a shorter URL prefix will never hide matching a longer one.
+            self._classes.sort(key=lambda e: len(e.url_prefix), reverse=True)
```

Comment:
This is probably fine performance-wise, but it feels like it might be a good place to use a `Trie` data structure. However, that would require adding a new dependency, so I'm not sure the work to do it is worth the minimal performance gain we'd see. Since this is very likely going to remain fairly small.